### PR TITLE
[529912, 529913] Add reboot option for SOTA

### PR DESCRIPTION
**Pull request by yengliong93**
_The PR review is to check for sustainability and correctness.  Sustainability is actually more business critical as correctness is largely tested into the code over time.   Its useful to keep in mind that SW often outlives the HW it was written for and engineers move from job to job so it is critical that code developed for Intel be supportable across many years.  It is up to the submitter and reviewer to look at the code from a perspective of what if we have to debug this 3 years from now after the author is no longer available and defect databases have been lost.  Yes, that happens all the time when we are working with time scales of more than 2 years.  When reviewing your code it is important to look at it from this perspective._

Author Mandatory (to be filled by PR Author/Submitter)
------------------------------------------------------
- Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.  
- Those checklist items which are not marked are considered as not applicable for the PR change.  
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the developers for Inner Source Development Model (ISDM) compliance

### PULL DESCRIPTION
This PR added a reboot option in OTA manifest and updated INBC to take the optional reboot option in OTA cmd.

### REFERENCES
Reference URL for issue tracking (JIRA/HSD/Github): **\<URL to be filled>**
- [x] **_Added label to the Pull Request following the template: ISDM\_\<Complexity>\*_** \
	Note-1: Depending on complexity of code changes, use the suitable word for complexity: Low/Medium/High \
	Example: PR for Slim boot loader project with medium complexity can have the label as: ISDM_Medium	
- [x] Added label to the Pull Request for easier discoverability and search
- [x] RTC or HSD number will be included in final merge. HSD must always be included if available.
- [x] Changelogs are updated (or N/A if not customer visible)
- NA inbm/log_changes.txt and inbm-vision/log_changes.txt are updated for potentially Validation-breaking log changes (or N/A if none)

### CODE MAINTAINABILITY
- [x] **_Commit Message meets guidelines as indicated in the URL\*_**
	- (https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md)
- [x] **_Every commit is a single defect fix and does not mix feature addition or changes\*_**
- [x] Added required new tests relevant to the changes
	- [ ] PR contains URL links to functional tests executed with the new tests 
- [x] Updated Documentation as relevant to the changes
- NA  Updated Build steps/commands changes as relevant
- NA  PR change contains code related to security
- NA  PR introduces changes that breaks compatibility with other modules (If YES, please provide description)
- NA  Specific instructions or information for code reviewers (If any):
- [x] Run 'go fmt' or autopep8 as applicable.
- [x] New/modified methods and functions should have type annotations on signatures as applicable
- [x] New/modified methods must have appropriate doc strings (language dependent)

### APPLICATION SPECIFIC
- [ ] Does PR change default config files under /etc? If so, will application still work after an upgrade that leaves /etc alone, like a Mender upgrade?
- [ ] Is cloud UI changed? If so, are cloud definition files updated?

Maintainer Mandatory (to be filled by PR Reviewer/Approving Maintainer)
-----------------------------------------------------------------------
- Maintainer who approves the Pull Request for merge is required to mark the checklist below as appropriate for the PR change reviewed as key proof of attestation indicating reasons for merge. 
- Those checklist items which are not marked are considered as not applicable for the PR change. 
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the maintainers for ISDM compliance.

### QUALITY CHECKS
- [ ] Architectural and Design Fit
- [ ] **_Quality of code (At least one should be checked as applicable)\*_**
	- [ ] Commit Message meets guidelines
	- [ ] PR changes adhere to industry practices and standards
	- [ ] Error and exception code paths implemented correctly
	- [ ] Code reviewed for domain or language specific anti-patterns
	- [ ] Code is adequately commented
	- [ ] Code copyright is correct
	- [ ] Confusing logic is explained in comments
	- [ ] Commit comment can be used to design a new test case for the changes
- [ ] **_Test coverage shows adequate coverage with required CI functional tests pass on all supported platforms\*_**
- [ ] **_Static code scan report shows zero critical issues\*_**
- [ ] Integration tests are passing

### CODE REVIEW IMPACT
- Summary of Defects Detected in Code Review: **\<%P1*xx,P2*xx,P3*xx,P4*xx%>** \
Note P1/P2/P3/P4 denotes severity of defects found (Showstopper/High/Medium/Low) and xx denotes number of defects found

### SECURITY CHECKS
Please check if your PR fulfills the following requirements:

- [ ] Follow best practices when handling primitive data types
- [ ] Configure minimal permissions when opening pipes and ports
- [ ] Check contents within input structures are valid before use
- [ ] All forms of input validated
- [ ] Avoid inter-process race conditions
- [ ] Error and exception handling implemented
- [ ] Defend against Canonical Representation Issues - Any paths utilized?
- [ ] Follow 'secure by default' - Any configuration values added
- [ ] Fail safe - Any failure scenarios?
- [ ] Clean up temporary files - Any temporary files being used?

# _Code must act as a teacher for future developers_

### CODE REVIEW

File: inbm/common-python-config/pyproject.toml [hunk 1]
```diff
@@ -41,4 +41,4 @@ max_tuple_unpack_length = 4
 [tool.flakeheaven.plugins]
 "flake8-annotations" = ["-*", "+ANN001", "+ANN2??", "+ANN301"]
 "wemake-python-styleguide" = ["-*", "+WPS2??", "-WPS220", "-WPS223", "-WPS222"] # WPS2?? is complexity; 220 is deep nesting; 223 is elifs; 222 is logic
-"flake8-bandit" = ["+*", "-S404", "-S603"]
+"flake8-bandit" = ["+*", "-S404", "-S603", "-E902"]
```
Comment (gblewis1) [inline inbm/common-python-config/pyproject.toml:44]: remove -E902

**Comment by tsirlapu** [comment]
@gblewis1 @nmgaston please review and let me know when it is good enough to be tested

File: inbm/integration-reloaded/test/sota/SOTA_UPDATE_SUCCESS_preboot.sh [hunk 2]
```diff
@@ -9,7 +9,7 @@ mkdir -p /var/log/sota_test
 echo "Triggering SOTA integration INB test 1: SOTA UPDATE SUCCESS"
 echo "<START> SOTA UPDATE SUCCESS" | systemd-cat
 
-inbc sota
+inbc sota --command update
```
Comment (gblewis1) [inline inbm/integration-reloaded/test/sota/SOTA_UPDATE_SUCCESS_preboot.sh:12]: there is no --command flag now

**Review by gblewis1 — CHANGES_REQUESTED** [review]
see comments

**Comment by gblewis1** [comment]
Some unit tests are failing. See https://tc01s-or.devtools.intel.com/viewLog.html?buildId=7136513&buildTypeId=TurtleCreek_AgentsTrtlPackagingIntegration&tab=buildLog

**Review by gblewis1 — APPROVED** [review]

File: inbc-program/README.md [hunk 3]
```diff
@@ -70,6 +70,7 @@ inbc fota {--uri, -u=URI}
    [--signature, -s SIGNATURE_STRING; default=None] 
    [--tooloptions, -to TOOL_OPTIONS]
    [--username, -un USERNAME] 
+   [--reboot; default=yes]
```
Comment (nmgaston) [inline inbc-program/README.md:73]: We need to define a shortcut command for this.  Maybe -rb?

File: inbc-program/README.md [hunk 4]
```diff
@@ -111,7 +112,8 @@ SOTA on Ubuntu is supported in 3 modes:
 inbc sota {--uri, -u=URI} 
    [--releasedate, -r RELEASE_DATE; default="2026-12-31"] 
    [--username, -un USERNAME]
-   [--mode, -m MODE; default="full", choices=["full", "no-download", "download-only"] ]
+   [--mode, -m MODE; default="full", choices=["full","no-download", "download-only"] ]
+   [--reboot; default=yes]
```
Comment (nmgaston) [inline inbc-program/README.md:116]: Same as above.  Update with a shortcut command for reboot.  -rb

File: inbc-program/README.md [hunk 5]
```diff
@@ -150,6 +152,7 @@ inbc pota {--fotauri, -fu=FOTA_URI}
    [--release_date, -sr SOTA_RELEASE_DATE; default="2026-12-31"] 
    [--fotasignature, -fs SIGNATURE_STRING] 
    [--username, -u USERNAME] 
+   [--reboot; default=yes]
```
Comment (nmgaston) [inline inbc-program/README.md:155]: Same as the other two.  Assign a shortcut command such as -rb

File: inbc-program/inbc/parser.py [hunk 6]
```diff
@@ -299,7 +305,8 @@ def sota(args) -> str:
                        "username",
                        "password",
                        "release_date",
-                       "path"
+                       "path",
+                       "deviceReboot"
```
Comment (nmgaston) [inline inbc-program/inbc/parser.py:309]: Why not just reboot instead of deviceReboot?  I think the device is implied and we can just say reboot.

Comment (gblewis1) [reply inbc-program/inbc/parser.py:309]: decided in IM to leave as deviceReboot

**Review by gblewis1 — CHANGES_REQUESTED** [review]

File: inbm/dispatcher-agent/dispatcher/sota/snapshot.py [hunk 7]
```diff
@@ -139,6 +142,13 @@ def take_snapshot(self) -> None:
 
                 dispatcher_state.write_dispatcher_state_to_state_file(state)
         except DispatcherException:
+            if check_filesystem_type() == EXT4:
```
[the remark this answers is not in the document]
Comment (gblewis1) [reply inbm/dispatcher-agent/dispatcher/sota/snapshot.py:145]: Third issue: if btrfs but snapper fails, this will still fail to update the log correctly.

**Review by gblewis1 — APPROVED** [review]
I made the snapshot more generic to non-ext4/non-btrfs filesystems and inbc sota --reboot no looks good on my system. @tsirlapu to test binaries at https://tc01s-or.devtools.intel.com/viewLog.html?buildId=7141834&buildTypeId=TurtleCreek_AgentsTrtlPackagingIntegration&tab=buildLog and then we can merge

**Pull request merged**